File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        '''
        There are two apporaches to 
        solve this problem

        The first one is iterative 
        the second one is Recusive
        
        for iterative
        Time complexity max(len(l1,l2))
        Space complexity max(len(11,l2))

        '''

        result=ListNode(None)
        temp=result
        L1=l1
        L2=l2
        carry=0

        while L1 or L2:
    
            current_sum=(L1.val if L1 else 0 )+(L2.val if L2 else 0)+carry
            
            carry=current_sum//10
            quotient=current_sum%10 if current_sum>9 else current_sum

            temp.next=ListNode(quotient)
            L1=L1.next if L1 else None
            L2=L2.next if L2 else None
            temp=temp.next
        if carry!=0:
            temp.next=ListNode(carry)
        return result.next
            
            
        

File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let carry=0
    let prevNode=new ListNode()
    let head=prevNode
   while(l1||l2 || carry ){
       let val1=0
       let val2=0
       if(l1){
           val1=l1.val
           l1=l1.next
       }
       if(l2){
           val2=l2.val
           l2=l2.next
       }
       let sum=val1+val2 +carry
       carry=Math.floor(sum/10)
       const currentNode=new ListNode((sum%10))
       prevNode.next=currentNode
       prevNode=currentNode
   }
   return head.next
};